README.md

```diff
@@ -2,20 +2,60 @@
 
 [![Build Status](https://travis-ci.org/bloomberg/clangmetatool.svg?branch=master)](https://travis-ci.org/bloomberg/clangmetatool)
 
-The point of this library is to establish a framework for reusing code
-in clang tools.
+## About clangmetatool
 
-Source control at [https://github.com/bloomberg/clangmetatool/](https://github.com/bloomberg/clangmetatool/)
+When we first started writing clang tools, we realized that there is a
+lot of life cycle management that we had to repeat. In some cases,
+people advocate the usage of global variables to manage the lifecycle
+of that data, but that makes code reuse across tools even harder.
 
-## Building
+Additionally, we also learned that when writing a tool, it will be
+benefitial if the code is split in two phases. First a data collection
+phase, and later a post-processing phase that actually performed the
+bulk of the logic of the tool.
 
-````bash
-mkdir build
-cd build
-cmake -DClang_DIR=/path/to/clang/cmake ..
-make
-make install
-````
+Essentially you will only need to write a class like:
+
+```
+class MyTool {
+private:
+  SomeDataCollector collector1;
+  SomeOtherDataCollector collector2;
+public:
+  MyTool(clang::CompilerInstance* ci, clang::ast_matchers::MatchFinder *f)
+   :collector1(ci, f), collector2(ci, f) {
+   // the individual collectors will register their callbacks in their
+   // constructor, the tool doesn't really need to do anything else here.
+  }
+  void postProcessing
+  (std::map<std::string, clang::tooling::Replacements> &replacementsMap) {
+   // use data from collector1 and collector2
+   // generate warnings and notices
+   // add replacements to replacementsMap
+  }
+};
+```
+
+And you can use the `clangmetatool::MetaToolFactory` combined with the
+`clangmetatool::Metatool` in your tool's main function:
+
+```
+int main(int argc, const char* argv[]) {
+  llvm::cl::OptionCategory MyToolCategory("my-tool options");
+  llvm::cl::extrahelp CommonHelp
+    (clang::tooling::CommonOptionsParser::HelpMessage);
+  clang::tooling::CommonOptionsParser
+    optionsParser(argc, argv, MyToolCategory);
+  clang::tooling::RefactoringTool tool(optionsParser.getCompilations(),
+                                       optionsParser.getSourcePathList());
+  clangmetatool::MetaToolFactory< clangmetatool::MetaTool<MyTool> >
+    raf(tool.getReplacements());
+  int r = tool.runAndSave(&raf);
+  return r;
+}
+```
+
```

Comment:
I curious as to whether this would be a good place to highlight why this is better, perhaps showing a minimal example of the old style?

Reply:
The example would be too verbose, but I added a paragraph talking about the downsides...